File: functions/src/quranSessions/sessionInProgressTransitionPolicy.ts
/** Q-SL-03: inProgress only when startsAt passed and join logged at/after start. */
export function shouldTransitionToInProgress(params: {
  startsAt: Date;
  now: Date;
  joinEventAtMs: number;
}): boolean {
  if (params.joinEventAtMs < params.startsAt.getTime()) {
    return false;
  }
  return params.now.getTime() >= params.startsAt.getTime();
}

File: packages/quran_sessions/lib/src/domain/policies/session_in_progress_transition_policy.dart
/// Rules for system `inProgress` transition (Q-SL-03).
abstract final class SessionInProgressTransitionPolicy {
  /// True when `startsAt` has passed and at least one join event was logged.
  static bool shouldTransitionToInProgress({
    required DateTime startsAt,
    required DateTime now,
    required bool hasJoinEventAtOrAfterStart,
  }) {
    if (!hasJoinEventAtOrAfterStart) return false;
    return !now.isBefore(startsAt);
  }
}
